File: src/BaseHmiTypes/Screens/Base/HmiScreenItemBase.cs
namespace BaseHmiTypes.Screens.Base;

using BaseHmiTypes.Common;
using BaseHmiTypes.Scripts;
using BaseHmiTypes.Screens.Widgets;

public abstract class HmiScreenItemBase : HmiScreenModelBase
{
    public string? Description { get; set; }

    public bool? IsWallpaper { get; set; }

    public bool? IsReferenceObject { get; set; }

    public HmiReferenceObjectSettings? ReferenceObject { get; set; }

    /// <summary>Gets the global-object parameter definitions declared by this base object.</summary>
    public IList<HmiScreenParameter> Parameters { get; } = new List<HmiScreenParameter>();

    public HmiScriptExposureMode? ScriptExposureMode { get; set; }

    public string? SourceFormat { get; set; }

    public byte[]? SourceData { get; set; }

    public Dictionary<string, string> SourceProperties { get; } = [];

    public HmiProperty<double> X { get; set; } = 0;

    public HmiProperty<double> Y { get; set; } = 0;

    public HmiProperty<double> Width { get; set; } = 0;

    public HmiProperty<double> Height { get; set; } = 0;

    public HmiProperty<bool> Visible { get; set; } = true;

    public HmiProperty<bool> Enabled { get; set; } = true;

    public HmiProperty<double>? Opacity { get; set; }

    public HmiProperty<double>? RotationAngle { get; set; }

    public HmiProperty<double>? RotationCenterX { get; set; }

    public HmiProperty<double>? RotationCenterY { get; set; }

    public HmiRangeAnimation? HorizontalPositionAnimation { get; set; }

    public HmiRangeAnimation? VerticalPositionAnimation { get; set; }

    public HmiRangeAnimation? WidthAnimation { get; set; }

    public HmiRangeAnimation? HeightAnimation { get; set; }

    public HmiRangeAnimation? RotationAnimation { get; set; }

    public HmiSliderAnimation? HorizontalSliderAnimation { get; set; }

    public HmiSliderAnimation? VerticalSliderAnimation { get; set; }

    public HmiAffineTransform? Transform { get; set; }

    public HmiProperty<int>? TabIndex { get; set; }

    public string? SecurityCode { get; set; }

    public HmiProperty<bool>? KeyNavigation { get; set; }

    public HmiProperty<HmiMultilingualText>? ToolTipText { get; set; }

    public HmiProperty<bool>? CanBeGrouped { get; set; }

    public HmiElectronicSignatureSettings? ElectronicSignature { get; set; }

    public HmiVariableConfirmationSettings? VariableConfirmation { get; set; }

    public HmiTouchAreaShape? TouchAreaShape { get; set; }

    public HmiConfirmationDialogSettings? TouchConfirmationDialog { get; set; }

    public IList<HmiEventBinding> Events { get; } = new List<HmiEventBinding>();
}

File: typescript/base-hmi-types/src/screens/base/HmiScreenItemBase.ts
import { HmiMultilingualText } from "../../common/HmiMultilingualText.js";
import { HmiScreenModelBase } from "./HmiScreenModelBase.js";
import { HmiProperty, staticProperty } from "./HmiProperty.js";
import { HmiEventBinding } from "../../scripts/HmiEventBinding.js";
import { HmiElectronicSignatureSettings } from "./HmiElectronicSignatureSettings.js";
import { HmiScriptExposureMode } from "./HmiScriptExposureMode.js";
import { HmiReferenceObjectSettings } from "./HmiReferenceObjectSettings.js";
import { HmiVariableConfirmationSettings } from "./HmiVariableConfirmationSettings.js";
import { HmiAffineTransform } from "./HmiAffineTransform.js";
import { HmiTouchAreaShape } from "./HmiTouchAreaShape.js";
import { HmiConfirmationDialogSettings } from "../widgets/HmiConfirmationDialogSettings.js";
import { HmiRangeAnimation } from "./HmiRangeAnimation.js";
import { HmiSliderAnimation } from "./HmiSliderAnimation.js";
import { HmiScreenParameter } from "./HmiScreenParameter.js";
import { HmiObjectConnection } from "./HmiObjectConnection.js";

export abstract class HmiScreenItemBase extends HmiScreenModelBase {
  description?: string;
  isWallpaper?: boolean;
  isReferenceObject?: boolean;
  referenceObject?: HmiReferenceObjectSettings;
  /** Global-object parameter definitions declared by this base object. */
  readonly parameters: HmiScreenParameter[] = [];
  /** Source connections assigned directly to this object. */
  readonly connections: HmiObjectConnection[] = [];
  scriptExposureMode?: HmiScriptExposureMode;
  sourceFormat?: string;
  sourceData?: Uint8Array;
  readonly sourceProperties: Record<string, string> = {};
  x: HmiProperty<number> = staticProperty(0);
  y: HmiProperty<number> = staticProperty(0);
  width: HmiProperty<number> = staticProperty(0);
  height: HmiProperty<number> = staticProperty(0);
  visible: HmiProperty<boolean> = staticProperty(true);
  enabled: HmiProperty<boolean> = staticProperty(true);
  /** Source polarity for an optional enabled-state expression. */
  enabledWhenExpressionIsTrue?: boolean;
  opacity?: HmiProperty<number>;
  rotationAngle?: HmiProperty<number>;
  rotationCenterX?: HmiProperty<number>;
  rotationCenterY?: HmiProperty<number>;
  horizontalPositionAnimation?: HmiRangeAnimation;
  verticalPositionAnimation?: HmiRangeAnimation;
  widthAnimation?: HmiRangeAnimation;
  heightAnimation?: HmiRangeAnimation;
  rotationAnimation?: HmiRangeAnimation;
  horizontalSliderAnimation?: HmiSliderAnimation;
  verticalSliderAnimation?: HmiSliderAnimation;
  transform?: HmiAffineTransform;
  tabIndex?: HmiProperty<number>;
  securityCode?: string;
  keyNavigation?: HmiProperty<boolean>;
  toolTipText?: HmiProperty<HmiMultilingualText>;
  canBeGrouped?: HmiProperty<boolean>;
  electronicSignature?: HmiElectronicSignatureSettings;
  variableConfirmation?: HmiVariableConfirmationSettings;
  touchAreaShape?: HmiTouchAreaShape;
  touchConfirmationDialog?: HmiConfirmationDialogSettings;
  readonly events: HmiEventBinding[] = [];
}
